File: evals/behavioral-analysis/data/javascript/arbitrary-resource-read-write/path_traversal_directory_enumeration.js
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: Path Traversal with Directory Enumeration
//
// Joins user filename to a base dir without canonicalisation; reads and lists.

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";
import fs from "node:fs";
import path from "node:path";

const BASE_DIR = "/app/data";

class PathTraversalReader {
  static constructPath(filename) {
    return path.join(BASE_DIR, filename);
  }

  static readFileContent(filepath) {
    try {
      const content = fs.readFileSync(filepath, "utf8");
      return { path: filepath, size: content.length, content: content.slice(0, 1000) };
    } catch (e) {
      return { path: filepath, error: String(e) };
    }
  }

  static enumerateDirectory(dirpath) {
    const files = [];
    try {
      for (const item of fs.readdirSync(dirpath)) {
        const fullPath = path.join(dirpath, item);
        try {
          const isDir = fs.statSync(fullPath).isDirectory();
          const size = isDir ? 0 : fs.statSync(fullPath).size;
          files.push({ name: item, path: fullPath, is_directory: isDir, size });
        } catch {
          /* skip */
        }
      }
    } catch {
      /* skip */
    }
    return files;
  }
}

const server = new Server(
  { name: "arbitrary-resource-service-1", version: "1.0.0" },
  { capabilities: { tools: {} } },
);

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [{
    name: "read_file",
    description: "Read application file with optional directory enumeration.",
    inputSchema: {
      type: "object",
      properties: {
        filename: { type: "string" },
        enumerate_dir: { type: "boolean", default: false },
      },
      required: ["filename"],
    },
  }],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const { filename, enumerate_dir = false } = req.params.arguments;
  const reader = PathTraversalReader;
  const filepath = reader.constructPath(filename);
  const result = reader.readFileContent(filepath);
  if (enumerate_dir && !result.error) {
    const dirpath = path.dirname(filepath);
    const files = reader.enumerateDirectory(dirpath);
    return {
      content: [{
        type: "text",
        text: `File read: ${result.size ?? 0} bytes, ${files.length} files in directory`,
      }],
    };
  }
  return { content: [{ type: "text", text: `File read: ${result.size ?? 0} bytes` }] };
});

File: evals/behavioral-analysis/data/csharp/arbitrary-resource-read-write/PathTraversalDirectoryEnumeration.cs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: Path Traversal with Directory Enumeration

using System;
using System.ComponentModel;
using System.IO;
using ModelContextProtocol.Server;

namespace EvalSamples.ArbitraryResource;

[McpServerToolType]
public static class PathTraversalDirectoryEnumeration
{
    private const string BaseDir = "/app/data";

    private static string ConstructPath(string filename) =>
        Path.Combine(BaseDir, filename);

    private static (int size, string? err) ReadFileContent(string filepath)
    {
        try
        {
            var content = File.ReadAllText(filepath);
            var n = Math.Min(1000, content.Length);
            return (content.Length, null);
        }
        catch (Exception ex)
        {
            return (0, ex.Message);
        }
    }

    private static int EnumerateDirectoryCount(string dirpath)
    {
        try
        {
            return Directory.GetFileSystemEntries(dirpath).Length;
        }
        catch
        {
            return 0;
        }
    }

    [McpServerTool]
    [Description("Read application file with optional directory enumeration.")]
    public static string ReadFile(string filename, bool enumerate_dir = false)
    {
        var filepath = ConstructPath(filename);
        var (size, err) = ReadFileContent(filepath);

        if (enumerate_dir && err == null)
        {
            var dir = Path.GetDirectoryName(filepath) ?? "/";
            var cnt = EnumerateDirectoryCount(dir);
            return $"File read: {size} bytes, {cnt} files in directory";
        }

        return $"File read: {size} bytes";
    }
}
